Userland/Utilities/touch.cpp

```diff
@@ -21,31 +38,227 @@ static bool file_exists(String const& path)
+        if (!maybe_value.has_value())
+            err("invalid datetime format '{}' -- expected number at index {}", datetime, lexer.tell());
+        else
+            value = maybe_value.value();
+    };
+
+    lex_number(year, 4);
+    if (!lexer.consume_specific('-'))
+        err("invalid datetime format '{}' -- expected '-' after year", datetime);
+    lex_number(month, 2);
+    if (!lexer.consume_specific('-'))
+        err("invalid datetime format '{}' -- expected '-' after month", datetime);
+    lex_number(day, 2);
+
+    // Parse the time designator -- a single 'T' or ' ' according to POSIX.
+    if (!lexer.consume_specific('T') && !lexer.consume_specific(' '))
+        err("invalid datetime format '{}' -- expected 'T' or ' ' for time designator", datetime);
+
+    lex_number(hour, 2);
+    if (!lexer.consume_specific(':'))
+        err("invalid datetime format '{}' -- expected ':' after hour", datetime);
+    lex_number(minute, 2);
+    if (!lexer.consume_specific(':'))
+        err("invalid datetime format '{}' -- expected ':' after minute", datetime);
+    lex_number(second, 2);
+
+    millisecond = 0;
+    if (!lexer.is_eof()) {
+        if (lexer.consume_specific(',') || lexer.consume_specific('.')) {
+            auto fractional_second = lexer.consume_while(isdigit);
+            if (fractional_second.is_empty())
+                err("invalid datetime format '{}' -- expected floating seconds", datetime);
+            for (u8 i = 0; i < 3 && i < fractional_second.length(); ++i) {
+                unsigned n = fractional_second[i] - '0';
+                switch (i) {
+                case 0:
+                    millisecond += 100 * n;
+                    break;
+                case 1:
+                    millisecond += 10 * n;
+                    break;
+                case 2:
+                    millisecond += n;
+                    break;
+                default:
+                    VERIFY_NOT_REACHED();
+                }
+            }
+        }
+
+        time_zone = lexer.consume_all();
+        if (time_zone != "Z")
+            err("invalid datetime format '{}' -- failed to parse time zone", datetime);
+    }
+
+    if (validate_timestamp(year, month, day, hour, minute, second)) {
+        auto timestamp = AK::Time::from_timestamp(year, month, day, hour, minute, second, millisecond);
+        auto time = timestamp.to_timespec();
+        if (time_zone.is_empty() && TimeZone::system_time_zone() != "UTC") {
+            auto offset = TimeZone::get_time_zone_offset(TimeZone::system_time_zone(), timestamp);
+            if (offset.has_value())
+                time.tv_sec -= offset.value().seconds;
+            else
+                err("failed to get the system time zone");
+        }
+        atime = mtime = time;
+    } else {
+        err("invalid datetime format '{}'", datetime);
     }
-    perror("stat");
-    exit(1);
+}
+
+static void reference_time(String const& strref, timespec& atime, timespec& mtime)
```

Comment:
Use a StringView instead of a string reference.

Reply:
Also, call this something like "reference_path" because that's what it does.

Reply:
I renamed `strref` to `reference_path` -- like you wrote, it's a much better name. However, I didn't convert the argument to a StringView because `stat()` requires a C-style null-terminated string, but it doesn't seem like StringView supports that conversion.

Reply:
Is there no `System::stat`? Probably time for one.

Reply:
You're right -- `Core::System::stat()` exists. `referece_time()` takes a StringView and uses `Core::system::stat()` now.

In `touch.cpp`, I also deleted `file_exists()` -- which used regular `stat()`, not `Core::System::stat()` -- and replaced it with `Core::File::exists()`. Although `Core::File::exists()` uses regular `stat()` as well, so I can revert this change if need be.

Reply:
Another nitpick: is there a `Core::Stream::file::exists`? We're trying to get rid of `Core::File` right now.

Reply:
Oh, I wasn't aware of that.

`Core::Stream::File::exists()` didn't exist, but I created it. Its implementation is functionally equivalent to `Core::File::exists()` as far as I'm aware, but it uses `Core::System::stat()` instead of regular `stat()`.